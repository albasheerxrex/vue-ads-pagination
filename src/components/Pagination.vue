<template>
    <div
        v-if="totalPages > 0"
        class="m-2 flex text-xs px-0"
    >
        <div class="pr-2 leading-loose">
            Items
            {{ oneBasedStart }}
            tot
            {{ oneBasedEnd }}
            van de
            {{ totalItems }}
        </div>
        <div
            v-if="totalPages > 1"
            class="flex-grow text-right px-0 flex justify-end"
        >
            <Button
                v-for="(button, key) in buttons"
                :key="key"
                :class="buttonClasses(button)"
                :disabled="button.disabled"
                :title="button.title"
                v-html="button.html"
                @click.native="button.page === undefined || button.active ? null : click(button.page)"
            >
            </Button>
        </div>
    </div>
</template>

<script>
import '../assets/css/styles.css';
import Button from './Button';

export default {
    name: 'Pagination',
    components: {Button},
    props: {
        startPage: {
            type: Number,
            required: false,
            default: 0,
        },

        itemsPerPage: {
            type: Number,
            required: false,
            default: 10,
        },

        maxVisiblePages: {
            type: Number,
            required: false,
            default: 5,
        },

        totalItems: {
            type: Number,
            required: true,
        },

        loading: {
            type: Boolean,
            required: false,
            default: false,
        },
    },

    created () {
        if (this.startPage < 0) {
            throw new Error(
                'Start page cannot be under zero'
            );
        }

        if (this.startPage > this.totalPages) {
            throw new Error(
                'Start page cannot be greater than the total number of pages'
            );
        }
    },

    data () {
        return {
            page: this.startPage,
        };
    },

    computed: {
        range () {
            let start = this.page * this.itemsPerPage;
            let end = start + this.itemsPerPage;

            return {
                start,
                end,
            };
        },

        totalPages () {
            if (this.itemsPerPage === 0) {
                return 0;
            }

            return Math.ceil(this.totalItems / this.itemsPerPage);
        },

        oneBasedStart () {
            return this.range.start + 1;
        },

        oneBasedEnd () {
            let oneBasedEnd = this.range.end;

            return this.totalItems < oneBasedEnd ? this.totalItems : oneBasedEnd;
        },

        buttons () {
            let diff = Math.ceil((this.maxVisiblePages - 1) / 2);
            let start = this.page - diff;
            let end = start + this.maxVisiblePages;

            if (start < diff) {
                end -= (start - diff);
                start = start < 0 ? 0 : start;
            }

            if (end + diff > this.totalPages) {
                start -= end - this.totalPages + diff;
                end = end > this.totalPages ? this.totalPages : end;
            }

            let buttons = Array
                .from(new Array(end - start), (value, index) => start + index)
                .map(value => this.getButton(value));

            if (start > 1) {
                let button = start === 2
                    ? this.getButton(1)
                    : this.getButton(undefined, '...', '', true, false);

                buttons.unshift(button);
            }

            if (start !== 0) {
                buttons.unshift(this.getButton(0));
            }

            if (end < this.totalPages - 1) {
                let button = end === this.totalPages - 2
                    ? this.getButton(this.totalPages - 2)
                    : this.getButton(undefined, '...', '', true, false);

                buttons.push(button);
            }

            if (end !== this.totalPages) {
                buttons.push(this.getButton(this.totalPages - 1));
            }

            buttons.unshift(
                this.getButton(
                    this.page - 1,
                    '<i class="fa fa-angle-double-left"></i>',
                    'previous',
                    this.page === 0,
                    true
                )
            );

            buttons.push(
                this.getButton(
                    this.page + 1,
                    '<i class="fa fa-angle-double-right"></i>',
                    'next',
                    this.page === this.totalPages - 1,
                    true
                )
            );

            return buttons;
        },
    },

    methods: {
        getButton (
            page = undefined,
            html = undefined,
            title = undefined,
            disabled = false,
            showDisabled = true
        ) {
            let active = page === this.page;
            return {
                page: page,
                html: html || page + 1,
                active,
                title: disabled || active ? '' : title || page + 1 || '',
                disabled,
                showDisabled,
            };
        },

        buttonClasses (button) {
            return {
                'text-grey': button.disabled && button.showDisabled,
                'bg-grey-light': button.disabled && button.showDisabled,
                'cursor-default': button.disabled || button.active,
                'border-none': button.disabled && !button.showDisabled,
                'bg-tt-green': button.active,
                'text-white': button.active,
                'hover:bg-grey-lighter': !button.active && !button.disabled,
            };
        },

        click (page) {
            this.page = page;
            this.$emit('click', this.page, this.range);
        },
    },
};
</script>
