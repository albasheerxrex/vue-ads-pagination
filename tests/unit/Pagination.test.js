import { shallowMount } from '@vue/test-utils';
import Pagination from '@/components/Pagination.vue';

describe('Pagination', () => {
    let props;

    beforeEach(() => {
        const itemsPerPage = 10;
        const maxVisiblePages = 5;
        const startPage = 0;
        const totalItems = 150;
        const loading = false;
        props = {
            itemsPerPage,
            maxVisiblePages,
            startPage,
            totalItems,
            loading,
        };
    });

    it('renders the pagination', function () {
        const pagination = shallowMount(Pagination, {
            propsData: props,
        });

        expect(pagination.html()).toMatchSnapshot();
    });

    it('renders the pagination if the page is equal to the last page', function () {
        props.startPage = 14;

        const pagination = shallowMount(Pagination, {
            propsData: props,
        });

        expect(pagination.html()).toMatchSnapshot();
    });

    it('renders nothing with itemsPerPage is 0', function () {
        props.itemsPerPage = 0;

        const pagination = shallowMount(Pagination, {
            propsData: props,
        });

        expect(pagination.html()).toBe(undefined);
    });

    it('renders no pages with only one page', function () {
        props.itemsPerPage = 150;

        const pagination = shallowMount(Pagination, {
            propsData: props,
        });

        expect(pagination.html()).toMatchSnapshot();
    });

    it('throws an error with negative start page', function () {
        props.startPage = -1;
        props.totalItems = 1;

        expect(() => {
            shallowMount(Pagination, {
                propsData: props,
            });
        }).toThrow('Start page cannot be under zero');
    });

    it('throws an error if the page is greater than the max page', function () {
        props.startPage = 2;
        props.totalItems = 1;

        expect(() => {
            shallowMount(Pagination, {
                propsData: props,
            });
        }).toThrow('Start page cannot be greater than the total number of pages');
    });

    it('emits the current page and the range on a click', function () {
        const pagination = shallowMount(Pagination, {
            propsData: props,
        });
        pagination.vm.click(1);

        expect(pagination.html()).toMatchSnapshot();
        expect(pagination.emitted().click).toBeTruthy();
        expect(pagination.emitted().click[0]).toEqual([
            1,
            {
                start: 10,
                end: 20,
            },
        ]);
    });
});
