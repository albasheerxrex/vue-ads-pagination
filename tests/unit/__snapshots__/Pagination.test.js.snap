// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination emits the current page and the range on a click 1`] = `
<div class="m-2 flex text-xs px-0">
  <div class="pr-2 leading-loose">
    Items 11 tot 20 van de 150
  </div>
  <div class="flex-grow text-right px-0 flex justify-end">
    <button-stub title="previous" class="hover:bg-grey-lighter">
      <i class="fa fa-angle-double-left"></i>
    </button-stub>
    <button-stub title="1" class="hover:bg-grey-lighter">1</button-stub>
    <button-stub title="" class="cursor-default bg-tt-green text-white">2</button-stub>
    <button-stub title="3" class="hover:bg-grey-lighter">3</button-stub>
    <button-stub title="4" class="hover:bg-grey-lighter">4</button-stub>
    <button-stub title="5" class="hover:bg-grey-lighter">5</button-stub>
    <button-stub title="6" class="hover:bg-grey-lighter">6</button-stub>
    <button-stub title="7" class="hover:bg-grey-lighter">7</button-stub>
    <button-stub disabled="true" title="" class="cursor-default border-none">...</button-stub>
    <button-stub title="15" class="hover:bg-grey-lighter">15</button-stub>
    <button-stub title="next" class="hover:bg-grey-lighter">
      <i class="fa fa-angle-double-right"></i>
    </button-stub>
  </div>
</div>
`;

exports[`Pagination renders no pages with only one page 1`] = `
<div class="m-2 flex text-xs px-0">
  <div class="pr-2 leading-loose">
    Items 1 tot 150 van de 150
  </div>
  <!---->
</div>
`;

exports[`Pagination renders the pagination 1`] = `
<div class="m-2 flex text-xs px-0">
  <div class="pr-2 leading-loose">
    Items 1 tot 10 van de 150
  </div>
  <div class="flex-grow text-right px-0 flex justify-end">
    <button-stub disabled="true" title="" class="text-grey bg-grey-light cursor-default">
      <i class="fa fa-angle-double-left"></i>
    </button-stub>
    <button-stub title="" class="cursor-default bg-tt-green text-white">1</button-stub>
    <button-stub title="2" class="hover:bg-grey-lighter">2</button-stub>
    <button-stub title="3" class="hover:bg-grey-lighter">3</button-stub>
    <button-stub title="4" class="hover:bg-grey-lighter">4</button-stub>
    <button-stub title="5" class="hover:bg-grey-lighter">5</button-stub>
    <button-stub title="6" class="hover:bg-grey-lighter">6</button-stub>
    <button-stub title="7" class="hover:bg-grey-lighter">7</button-stub>
    <button-stub disabled="true" title="" class="cursor-default border-none">...</button-stub>
    <button-stub title="15" class="hover:bg-grey-lighter">15</button-stub>
    <button-stub title="next" class="hover:bg-grey-lighter">
      <i class="fa fa-angle-double-right"></i>
    </button-stub>
  </div>
</div>
`;

exports[`Pagination renders the pagination if the page is equal to the last page 1`] = `
<div class="m-2 flex text-xs px-0">
  <div class="pr-2 leading-loose">
    Items 141 tot 150 van de 150
  </div>
  <div class="flex-grow text-right px-0 flex justify-end">
    <button-stub title="previous" class="hover:bg-grey-lighter">
      <i class="fa fa-angle-double-left"></i>
    </button-stub>
    <button-stub title="1" class="hover:bg-grey-lighter">1</button-stub>
    <button-stub disabled="true" title="" class="cursor-default border-none">...</button-stub>
    <button-stub title="9" class="hover:bg-grey-lighter">9</button-stub>
    <button-stub title="10" class="hover:bg-grey-lighter">10</button-stub>
    <button-stub title="11" class="hover:bg-grey-lighter">11</button-stub>
    <button-stub title="12" class="hover:bg-grey-lighter">12</button-stub>
    <button-stub title="13" class="hover:bg-grey-lighter">13</button-stub>
    <button-stub title="14" class="hover:bg-grey-lighter">14</button-stub>
    <button-stub title="" class="cursor-default bg-tt-green text-white">15</button-stub>
    <button-stub disabled="true" title="" class="text-grey bg-grey-light cursor-default">
      <i class="fa fa-angle-double-right"></i>
    </button-stub>
  </div>
</div>
`;
